File: src/services/bedrock.js
// services/bedrock.js — AWS Bedrock Runtime emulator
//
// Path-routed (the bedrock-runtime SDK uses REST-JSON, not X-Amz-Target):
//   POST /model/<modelId>/invoke
//   POST /model/<modelId>/invoke-with-response-stream   (vnd.amazon.eventstream)
//   POST /model/<modelId>/converse
//   POST /model/<modelId>/converse-stream               (vnd.amazon.eventstream)
//   POST /guardrail/<id>/version/<v>/apply              (stub)
//
// Responses are CANNED and CONFIGURABLE via store.bedrock + the
// /mockcloud/bedrock control plane (src/routes/bedrock.js): callers register
// rules that match on model + prompt substring and return a fixed completion
// or inject a fault. This makes Bedrock-backed code testable offline without a
// real model. Scope: InvokeModel + Converse (+ streaming) + a guardrail stub —
// not the full Bedrock surface.
import { store, randomId } from '../store.js';
import { jsonResponse, getParsedBody } from '../middleware/response.js';

const PATH_RE = /^\/model\/(.+)\/(invoke-with-response-stream|invoke|converse-stream|converse)$/;

export function handler(req, res) {
  const path = new URL(req.url, 'http://x').pathname;

  if (path.startsWith('/guardrail/')) return applyGuardrail(req, res);

  const m = PATH_RE.exec(path);
  if (!m) return jsonResponse(res, 404, { __type: 'ResourceNotFoundException', message: `Unknown Bedrock path: ${path}` });
  const modelId = decodeURIComponent(m[1]);
  const op = m[2];
  const body = getParsedBody(req);
  const prompt = extractPrompt(body);

  const resolved = resolveRule(modelId, prompt);
  recordInvocation(modelId, op, prompt, resolved);

  // Fault injection — return the configured error before producing output.
  if (resolved.fault) {
    const f = resolved.fault;
    return jsonResponse(res, f.statusCode || 400, { __type: f.type || 'ThrottlingException', message: f.message || 'Mock Bedrock fault' });
  }

  const text = resolved.response;
  const inputTokens = approxTokens(prompt);
  const outputTokens = approxTokens(text);

  switch (op) {
    case 'invoke':         return invokeModel(res, modelId, text, inputTokens, outputTokens);
    case 'converse':       return converse(res, modelId, text, inputTokens, outputTokens);
    case 'invoke-with-response-stream': return invokeModelStream(res, text, inputTokens, outputTokens);
    case 'converse-stream':             return converseStream(res, text, inputTokens, outputTokens);
  }
}

// ── Non-streaming ───────────────────────────────────────────────────────────
function invokeModel(res, modelId, text, inputTokens, outputTokens) {
  // bedrock-runtime InvokeModel returns the raw model body. Shape it per family.
  const body = modelId.startsWith('anthropic.')
    ? {
        id: `msg_${randomId(24)}`, type: 'message', role: 'assistant', model: modelId,
        content: [{ type: 'text', text }], stop_reason: 'end_turn', stop_sequence: null,
        usage: { input_tokens: inputTokens, output_tokens: outputTokens },
      }
    : modelId.startsWith('amazon.titan')
    ? { inputTextTokenCount: inputTokens, results: [{ tokenCount: outputTokens, outputText: text, completionReason: 'FINISH' }] }
    : { outputText: text, completionReason: 'FINISH' };
  return jsonResponse(res, 200, body);
}

function converse(res, modelId, text, inputTokens, outputTokens) {
  return jsonResponse(res, 200, {
    output: { message: { role: 'assistant', content: [{ text }] } },
    stopReason: 'end_turn',
    usage: { inputTokens, outputTokens, totalTokens: inputTokens + outputTokens },
    metrics: { latencyMs: 5 },
  });
}

// ── Streaming (vnd.amazon.eventstream) ───────────────────────────────────────
function startStream(res) {
  res.writeHead(200, { 'Content-Type': 'application/vnd.amazon.eventstream', 'x-amzn-RequestId': randomId(32) });
}

// InvokeModelWithResponseStream: a series of `chunk` events whose payload is
// { bytes: base64(<model streaming chunk JSON>) } — Anthropic message events.
function invokeModelStream(res, text, inputTokens, outputTokens) {
  startStream(res);
  const chunk = obj => res.write(eventFrame('chunk', { bytes: Buffer.from(JSON.stringify(obj)).toString('base64') }));
  chunk({ type: 'message_start', message: { role: 'assistant', usage: { input_tokens: inputTokens, output_tokens: 0 } } });
  chunk({ type: 'content_block_start', index: 0, content_block: { type: 'text', text: '' } });
  for (const piece of splitForStream(text)) {
    chunk({ type: 'content_block_delta', index: 0, delta: { type: 'text_delta', text: piece } });
  }
  chunk({ type: 'content_block_stop', index: 0 });
  chunk({ type: 'message_delta', delta: { stop_reason: 'end_turn' }, usage: { output_tokens: outputTokens } });
  chunk({ type: 'message_stop' });
  res.end();
}

// ConverseStream: typed events (messageStart / contentBlockDelta / ... / metadata).
function converseStream(res, text, inputTokens, outputTokens) {
  startStream(res);
  res.write(eventFrame('messageStart', { role: 'assistant' }));
  for (const piece of splitForStream(text)) {
    res.write(eventFrame('contentBlockDelta', { contentBlockIndex: 0, delta: { text: piece } }));
  }
  res.write(eventFrame('contentBlockStop', { contentBlockIndex: 0 }));
  res.write(eventFrame('messageStop', { stopReason: 'end_turn' }));
  res.write(eventFrame('metadata', {
    usage: { inputTokens, outputTokens, totalTokens: inputTokens + outputTokens },
    metrics: { latencyMs: 5 },
  }));
  res.end();
}

// Guardrail apply — minimal stub: never intervenes.
function applyGuardrail(req, res) {
  const body = getParsedBody(req);
  const outputs = body.content?.map(c => ({ text: c.text?.text ?? '' })) || [];
  return jsonResponse(res, 200, { usage: {}, action: 'NONE', outputs, assessments: [] });
}

// ── Rule resolution / config ──────────────────────────────────────────────
// A rule matches when (no model || modelId matches the model glob) AND
// (no promptContains || prompt includes it). First match wins; falls back to
// the configured defaultResponse.
function resolveRule(modelId, prompt) {
  const cfg = store.bedrock;
  for (const rule of cfg.rules || []) {
    if (rule.model && !globMatch(rule.model, modelId)) continue;
    if (rule.promptContains && !(prompt || '').includes(rule.promptContains)) continue;
    return { response: rule.response ?? cfg.defaultResponse, fault: rule.fault || null };
  }
  return { response: cfg.defaultResponse, fault: null };
}

function recordInvocation(modelId, op, prompt, resolved) {
  store.bedrock.invocations.unshift({
    id: randomId(16), t: Date.now(), modelId, op,
    prompt: (prompt || '').slice(0, 500),
    faulted: !!resolved.fault,
  });
  if (store.bedrock.invocations.length > 200) store.bedrock.invocations.pop();
  store.addTrail({ method: 'POST', path: `/bedrock/${op}/${modelId}`, status: resolved.fault ? (resolved.fault.statusCode || 400) : 200, latency: 5 });
}

// ── Helpers ────────────────────────────────────────────────────────────────
// Pull the user prompt out of whatever request shape arrived.
function extractPrompt(body) {
  if (!body || typeof body !== 'object') return '';
  if (Array.isArray(body.messages) && body.messages.length) {
    const last = body.messages[body.messages.length - 1];
    return contentToText(last?.content);
  }
  return body.inputText || body.prompt || '';
}

function contentToText(content) {
  if (typeof content === 'string') return content;
  if (Array.isArray(content)) return content.map(c => (typeof c === 'string' ? c : c?.text ?? '')).join(' ').trim();
  return '';
}

// Rough token estimate (~4 chars/token) — enough for plausible usage numbers.
function approxTokens(s) { return Math.max(1, Math.ceil((s || '').length / 4)); }

// Split a canned response into a handful of streaming pieces (by word).
function splitForStream(text) {
  const words = String(text || '').split(/(\s+)/).filter(Boolean);
  return words.length ? words : [String(text || '')];
}

// Simple `*` glob: 'anthropic.*' matches 'anthropic.claude-...'; exact otherwise.
function globMatch(pattern, value) {
  if (pattern === value) return true;
  if (!pattern.includes('*')) return false;
  const re = new RegExp('^' + pattern.split('*').map(escapeRe).join('.*') + '$');
  return re.test(value);
}
function escapeRe(s) { return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'); }

// ── vnd.amazon.eventstream framing ──────────────────────────────────────────
// Frame: [totalLen u32][headerLen u32][preludeCRC u32][headers][payload][msgCRC u32]
// Header: [nameLen u8][name][valueType u8=7][valueLen u16][value]
function eventFrame(eventType, payloadObj) {
  return encodeMessage(
    { ':message-type': 'event', ':event-type': eventType, ':content-type': 'application/json' },
    Buffer.from(JSON.stringify(payloadObj)),
  );
}

function encodeMessage(headers, payloadBuf) {
  const headerBuf = Buffer.concat(Object.entries(headers).map(([k, v]) => encodeHeader(k, v)));
  const totalLen = 4 + 4 + 4 + headerBuf.length + payloadBuf.length + 4;
  const msg = Buffer.alloc(totalLen);
  let o = 0;
  msg.writeUInt32BE(totalLen, o); o += 4;
  msg.writeUInt32BE(headerBuf.length, o); o += 4;
  msg.writeUInt32BE(crc32(msg.subarray(0, 8)), o); o += 4;
  headerBuf.copy(msg, o); o += headerBuf.length;
  payloadBuf.copy(msg, o); o += payloadBuf.length;
  msg.writeUInt32BE(crc32(msg.subarray(0, o)), o);
  return msg;
}

function encodeHeader(name, value) {
  const nameBuf = Buffer.from(name, 'utf8');
  const valBuf = Buffer.from(String(value), 'utf8');
  const buf = Buffer.alloc(1 + nameBuf.length + 1 + 2 + valBuf.length);
  let o = 0;
  buf.writeUInt8(nameBuf.length, o); o += 1;
  nameBuf.copy(buf, o); o += nameBuf.length;
  buf.writeUInt8(7, o); o += 1;                 // value type 7 = UTF-8 string
  buf.writeUInt16BE(valBuf.length, o); o += 2;
  valBuf.copy(buf, o);
  return buf;
}

// CRC32 (IEEE 802.3) — Node's zlib.crc32 only exists on v22+, so table it here.
const CRC_TABLE = (() => {
  const t = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = (c & 1) ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1);
    t[n] = c >>> 0;
  }
  return t;
})();
function crc32(buf) {
  let c = 0xFFFFFFFF;
  for (let i = 0; i < buf.length; i++) c = CRC_TABLE[(c ^ buf[i]) & 0xFF] ^ (c >>> 8);
  return (c ^ 0xFFFFFFFF) >>> 0;
}

File: internal/services/bedrock/bedrock.go
// Package bedrock — port of src/services/bedrock.js. Path-routed REST-JSON
// (POST /model/<id>/{invoke,invoke-with-response-stream,converse,
// converse-stream}, /guardrail/*). Responses are canned and configurable via
// store.bedrock + the /mockcloud/bedrock control plane; streaming uses the
// vnd.amazon.eventstream binary framing.
package bedrock

import (
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"hash/crc32"
	"net/http"
	"net/url"
	"regexp"
	"strings"

	"github.com/mockcloud/mockcloud/internal/httpapi"
	"github.com/mockcloud/mockcloud/internal/protocol/respond"
	"github.com/mockcloud/mockcloud/internal/state"
	"github.com/mockcloud/mockcloud/internal/store"
)

type Service struct{ st *store.Store }

func New(st *store.Store) *Service { return &Service{st: st} }

var pathRe = regexp.MustCompile(`^/model/(.+)/(invoke-with-response-stream|invoke|converse-stream|converse)$`)

func (svc *Service) Handler(w http.ResponseWriter, r *httpapi.Request) {
	path := r.URL.EscapedPath()
	if strings.HasPrefix(path, "/guardrail/") {
		svc.applyGuardrail(w, r)
		return
	}
	m := pathRe.FindStringSubmatch(path)
	if m == nil {
		respond.JSON(w, 404, map[string]any{"__type": "ResourceNotFoundException", "message": "Unknown Bedrock path: " + path})
		return
	}
	modelID, err := url.PathUnescape(m[1])
	if err != nil {
		modelID = m[1]
	}
	op := m[2]
	body := r.JSONBody()
	prompt := extractPrompt(body)

	respText, fault := svc.resolveRule(modelID, prompt)
	svc.recordInvocation(modelID, op, prompt, fault)

	if fault != nil {
		status := intOr(fault["statusCode"], 400)
		respond.JSON(w, status, map[string]any{
			"__type":  strOr(fault["type"], "ThrottlingException"),
			"message": strOr(fault["message"], "Mock Bedrock fault"),
		})
		return
	}

	inputTokens := approxTokens(prompt)
	outputTokens := approxTokens(respText)

	switch op {
	case "invoke":
		svc.invokeModel(w, modelID, respText, inputTokens, outputTokens)
	case "converse":
		svc.converse(w, respText, inputTokens, outputTokens)
	case "invoke-with-response-stream":
		svc.invokeModelStream(w, respText, inputTokens, outputTokens)
	case "converse-stream":
		svc.converseStream(w, respText, inputTokens, outputTokens)
	}
}

// ── Non-streaming ───────────────────────────────────────────────────────────

func (svc *Service) invokeModel(w http.ResponseWriter, modelID, text string, inputTokens, outputTokens int) {
	var body map[string]any
	switch {
	case strings.HasPrefix(modelID, "anthropic."):
		body = map[string]any{
			"id": "msg_" + state.RandomID(24), "type": "message", "role": "assistant", "model": modelID,
			"content":     []any{map[string]any{"type": "text", "text": text}},
			"stop_reason": "end_turn", "stop_sequence": nil,
			"usage": map[string]any{"input_tokens": inputTokens, "output_tokens": outputTokens},
		}
	case strings.HasPrefix(modelID, "amazon.titan"):
		body = map[string]any{
			"inputTextTokenCount": inputTokens,
			"results":             []any{map[string]any{"tokenCount": outputTokens, "outputText": text, "completionReason": "FINISH"}},
		}
	default:
		body = map[string]any{"outputText": text, "completionReason": "FINISH"}
	}
	respond.JSON(w, 200, body)
}

func (svc *Service) converse(w http.ResponseWriter, text string, inputTokens, outputTokens int) {
	respond.JSON(w, 200, map[string]any{
		"output":     map[string]any{"message": map[string]any{"role": "assistant", "content": []any{map[string]any{"text": text}}}},
		"stopReason": "end_turn",
		"usage":      map[string]any{"inputTokens": inputTokens, "outputTokens": outputTokens, "totalTokens": inputTokens + outputTokens},
		"metrics":    map[string]any{"latencyMs": 5},
	})
}

// ── Streaming (vnd.amazon.eventstream) ───────────────────────────────────────

func startStream(w http.ResponseWriter) {
	w.Header().Set("Content-Type", "application/vnd.amazon.eventstream")
	w.Header().Set("x-amzn-RequestId", state.RandomID(32))
	w.WriteHeader(200)
}

func flush(w http.ResponseWriter) {
	if f, ok := w.(http.Flusher); ok {
		f.Flush()
	}
}

func (svc *Service) invokeModelStream(w http.ResponseWriter, text string, inputTokens, outputTokens int) {
	startStream(w)
	chunk := func(obj map[string]any) {
		payload := map[string]any{"bytes": base64.StdEncoding.EncodeToString(respond.Marshal(obj))}
		_, _ = w.Write(eventFrame("chunk", payload))
		flush(w)
	}
	chunk(map[string]any{"type": "message_start", "message": map[string]any{"role": "assistant", "usage": map[string]any{"input_tokens": inputTokens, "output_tokens": 0}}})
	chunk(map[string]any{"type": "content_block_start", "index": 0, "content_block": map[string]any{"type": "text", "text": ""}})
	for _, piece := range splitForStream(text) {
		chunk(map[string]any{"type": "content_block_delta", "index": 0, "delta": map[string]any{"type": "text_delta", "text": piece}})
	}
	chunk(map[string]any{"type": "content_block_stop", "index": 0})
	chunk(map[string]any{"type": "message_delta", "delta": map[string]any{"stop_reason": "end_turn"}, "usage": map[string]any{"output_tokens": outputTokens}})
	chunk(map[string]any{"type": "message_stop"})
}

func (svc *Service) converseStream(w http.ResponseWriter, text string, inputTokens, outputTokens int) {
	startStream(w)
	write := func(eventType string, payload map[string]any) {
		_, _ = w.Write(eventFrame(eventType, payload))
		flush(w)
	}
	write("messageStart", map[string]any{"role": "assistant"})
	for _, piece := range splitForStream(text) {
		write("contentBlockDelta", map[string]any{"contentBlockIndex": 0, "delta": map[string]any{"text": piece}})
	}
	write("contentBlockStop", map[string]any{"contentBlockIndex": 0})
	write("messageStop", map[string]any{"stopReason": "end_turn"})
	write("metadata", map[string]any{
		"usage":   map[string]any{"inputTokens": inputTokens, "outputTokens": outputTokens, "totalTokens": inputTokens + outputTokens},
		"metrics": map[string]any{"latencyMs": 5},
	})
}

func (svc *Service) applyGuardrail(w http.ResponseWriter, r *httpapi.Request) {
	body := r.JSONBody()
	outputs := []any{}
	if content, ok := body["content"].([]any); ok {
		for _, c := range content {
			cm, _ := c.(map[string]any)
			text := ""
			if inner, ok := cm["text"].(map[string]any); ok {
				text, _ = inner["text"].(string)
			}
			outputs = append(outputs, map[string]any{"text": text})
		}
	}
	respond.JSON(w, 200, map[string]any{"usage": map[string]any{}, "action": "NONE", "outputs": outputs, "assessments": []any{}})
}

// ── Rule resolution / config ─────────────────────────────────────────────────

func (svc *Service) resolveRule(modelID, prompt string) (string, map[string]any) {
	var text string
	var fault map[string]any
	svc.st.With(func(s *state.State) {
		def := s.Bedrock.DefaultResponse
		for _, rule := range s.Bedrock.Rules {
			if model, ok := rule["model"].(string); ok && model != "" && !globMatch(model, modelID) {
				continue
			}
			if pc, ok := rule["promptContains"].(string); ok && pc != "" && !strings.Contains(prompt, pc) {
				continue
			}
			text = def
			if r, ok := rule["response"].(string); ok {
				text = r
			}
			if f, ok := rule["fault"].(map[string]any); ok {
				fault = f
			}
			return
		}
		text = def
	})
	return text, fault
}

func (svc *Service) recordInvocation(modelID, op, prompt string, fault map[string]any) {
	svc.st.With(func(s *state.State) {
		if len(prompt) > 500 {
			prompt = prompt[:500]
		}
		inv := map[string]any{
			"id": state.RandomID(16), "t": state.NowMs(), "modelId": modelID, "op": op,
			"prompt": prompt, "faulted": fault != nil,
		}
		s.Bedrock.Invocations = append([]map[string]any{inv}, s.Bedrock.Invocations...)
		if len(s.Bedrock.Invocations) > 200 {
			s.Bedrock.Invocations = s.Bedrock.Invocations[:200]
		}
		status := 200
		if fault != nil {
			status = intOr(fault["statusCode"], 400)
		}
		s.AddTrail(map[string]any{"method": "POST", "path": "/bedrock/" + op + "/" + modelID, "status": status, "latency": 5})
	})
}

// RegisterUIRoutes — the /mockcloud/bedrock control plane (src/routes/bedrock.js).
func (svc *Service) RegisterUIRoutes(add func(method, pattern string, h func(http.ResponseWriter, *httpapi.Request))) {
	add("GET", "/mockcloud/bedrock", func(w http.ResponseWriter, r *httpapi.Request) {
		var out map[string]any
		svc.st.With(func(s *state.State) {
			invs := s.Bedrock.Invocations
			if len(invs) > 50 {
				invs = invs[:50]
			}
			out = map[string]any{"defaultResponse": s.Bedrock.DefaultResponse, "rules": s.Bedrock.Rules, "invocations": invs}
		})
		respond.JSON(w, 200, out)
	})
	add("PUT", "/mockcloud/bedrock", func(w http.ResponseWriter, r *httpapi.Request) {
		b := r.ParsedBody
		var def string
		svc.st.With(func(s *state.State) {
			if v, ok := b["defaultResponse"].(string); ok {
				s.Bedrock.DefaultResponse = v
			}
			def = s.Bedrock.DefaultResponse
		})
		respond.JSON(w, 200, map[string]any{"defaultResponse": def})
	})
	add("POST", "/mockcloud/bedrock/rules", func(w http.ResponseWriter, r *httpapi.Request) {
		b := r.ParsedBody
		rule := map[string]any{
			"id":             state.RandomID(8),
			"model":          orNil(b["model"]),
			"promptContains": orNil(b["promptContains"]),
			"fault":          orNil(b["fault"]),
		}
		if resp, ok := b["response"].(string); ok {
			rule["response"] = resp
		}
		svc.st.With(func(s *state.State) { s.Bedrock.Rules = append(s.Bedrock.Rules, rule) })
		respond.JSON(w, 201, rule)
	})
	add("DELETE", "/mockcloud/bedrock/rules", func(w http.ResponseWriter, r *httpapi.Request) {
		svc.st.With(func(s *state.State) { s.Bedrock.Rules = []map[string]any{} })
		respond.JSON(w, 200, map[string]any{"cleared": true})
	})
	add("DELETE", "/mockcloud/bedrock", func(w http.ResponseWriter, r *httpapi.Request) {
		svc.st.With(func(s *state.State) { s.Reset("bedrock") })
		respond.JSON(w, 200, map[string]any{"reset": true})
	})
}

// ── Helpers ───────────────────────────────────────────────────────────────────

func extractPrompt(body map[string]any) string {
	if body == nil {
		return ""
	}
	if msgs, ok := body["messages"].([]any); ok && len(msgs) > 0 {
		last, _ := msgs[len(msgs)-1].(map[string]any)
		if last != nil {
			return contentToText(last["content"])
		}
	}
	if v, ok := body["inputText"].(string); ok {
		return v
	}
	if v, ok := body["prompt"].(string); ok {
		return v
	}
	return ""
}

func contentToText(content any) string {
	switch c := content.(type) {
	case string:
		return c
	case []any:
		var parts []string
		for _, item := range c {
			if s, ok := item.(string); ok {
				parts = append(parts, s)
			} else if m, ok := item.(map[string]any); ok {
				if t, ok := m["text"].(string); ok {
					parts = append(parts, t)
				} else {
					parts = append(parts, "")
				}
			}
		}
		return strings.TrimSpace(strings.Join(parts, " "))
	}
	return ""
}

func approxTokens(s string) int {
	n := (len(s) + 3) / 4
	if n < 1 {
		return 1
	}
	return n
}

// splitForStream — split by whitespace runs, keeping separators (JS
// split(/(\s+)/).filter(Boolean)); empty text → [text].
var wsSplitRe = regexp.MustCompile(`(\s+)`)

func splitForStream(text string) []string {
	if text == "" {
		return []string{""}
	}
	var out []string
	last := 0
	for _, loc := range wsSplitRe.FindAllStringIndex(text, -1) {
		if loc[0] > last {
			out = append(out, text[last:loc[0]]) // word
		}
		out = append(out, text[loc[0]:loc[1]]) // separator
		last = loc[1]
	}
	if last < len(text) {
		out = append(out, text[last:])
	}
	if len(out) == 0 {
		return []string{text}
	}
	return out
}

func globMatch(pattern, value string) bool {
	if pattern == value {
		return true
	}
	if !strings.Contains(pattern, "*") {
		return false
	}
	parts := strings.Split(pattern, "*")
	for i, p := range parts {
		parts[i] = regexp.QuoteMeta(p)
	}
	re, err := regexp.Compile("^" + strings.Join(parts, ".*") + "$")
	return err == nil && re.MatchString(value)
}

func intOr(v any, def int) int {
	switch n := v.(type) {
	case float64:
		return int(n)
	case json.Number:
		if i, err := n.Int64(); err == nil {
			return int(i)
		}
	}
	return def
}

func strOr(v any, def string) string {
	if s, ok := v.(string); ok {
		return s
	}
	return def
}

func orNil(v any) any {
	if v == nil {
		return nil
	}
	// Node's `|| null`: empty string is falsy → null.
	if s, ok := v.(string); ok && s == "" {
		return nil
	}
	return v
}

// ── vnd.amazon.eventstream framing ───────────────────────────────────────────
// Frame: [totalLen u32][headerLen u32][preludeCRC u32][headers][payload][msgCRC u32]
// Header: [nameLen u8][name][valueType u8=7][valueLen u16][value]

func eventFrame(eventType string, payloadObj map[string]any) []byte {
	headers := []hdr{
		{":message-type", "event"},
		{":event-type", eventType},
		{":content-type", "application/json"},
	}
	return encodeMessage(headers, respond.Marshal(payloadObj))
}

type hdr struct{ name, value string }

func encodeMessage(headers []hdr, payload []byte) []byte {
	var headerBuf []byte
	for _, h := range headers {
		headerBuf = append(headerBuf, encodeHeader(h.name, h.value)...)
	}
	totalLen := 4 + 4 + 4 + len(headerBuf) + len(payload) + 4
	msg := make([]byte, totalLen)
	o := 0
	binary.BigEndian.PutUint32(msg[o:], uint32(totalLen))
	o += 4
	binary.BigEndian.PutUint32(msg[o:], uint32(len(headerBuf)))
	o += 4
	binary.BigEndian.PutUint32(msg[o:], crc32.ChecksumIEEE(msg[0:8])) // prelude CRC
	o += 4
	copy(msg[o:], headerBuf)
	o += len(headerBuf)
	copy(msg[o:], payload)
	o += len(payload)
	binary.BigEndian.PutUint32(msg[o:], crc32.ChecksumIEEE(msg[0:o])) // message CRC
	return msg
}

func encodeHeader(name, value string) []byte {
	nameB := []byte(name)
	valB := []byte(value)
	buf := make([]byte, 1+len(nameB)+1+2+len(valB))
	o := 0
	buf[o] = byte(len(nameB))
	o++
	copy(buf[o:], nameB)
	o += len(nameB)
	buf[o] = 7 // value type 7 = UTF-8 string
	o++
	binary.BigEndian.PutUint16(buf[o:], uint16(len(valB)))
	o += 2
	copy(buf[o:], valB)
	return buf
}
